### PR TITLE
Add automember module

### PR DESCRIPTION
**Pull request by jake2184**
I've extended & updated an automember module, first proposed in #392. That PR looked active for a couple of weeks, but silent for months. I've forked it, rebased on master, and addressed a few of the issues highlighted in the PR.

I can't tag the previous author, Mark Hahl / @Wolskie. I haven't changed any of the author notes or copyright.

### CODE REVIEW

**Comment by jake2184** [comment]
oh, the tag did work! Just didn't autocomplete.

**Comment by rjeffman** [comment]
Hello @jake2184,

The task failing is due to differences in string handling in Python 2 and 3. To fix this issue you should add the following to the imports:

```python
from ansible.module_utils._text import to_text
```

And then, when you need the string, you use `to_text(data)`

Thank you for you contribution!

**Comment by jake2184** [comment]
I seem to be getting errors now that I'm fairly sure are unrelated to any of my changes: 
```
> {"ansible_facts": {"discovered_interpreter_python": "/usr/libexec/platform-python"}, "changed": false, "msg": "Kerberos authentication failed: kinit: Cannot read password while getting initial credentials\n"
```

This is during facts collection? Is there anything I need to setup (rebase on master?) or change to get these working?

**Comment by rjeffman** [comment]
That is really weird... I'll look into that in the next days, as I really don't have the time now.

File: playbooks/automember/automember-absent.yml
Comment (rjeffman) [inline playbooks/automember/automember-absent.yml:13]: Please, separate this playbook in two, one for `group`, one for `hostgroup`.

File: playbooks/automember/automember-present.yml
Comment (rjeffman) [inline playbooks/automember/automember-present.yml:13]: Please, separate this playbook in two, one for group, one for hostgroup.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:25]: Either format like `black`, or (better yet) add the list of imports in the least number of lines possible.

If you start the list in the second line, with a `tab`, it will work:
```python
from package import (
    method01, method02, method03, method04,
    methdo05
)
```

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:230]: Since `type` doesn't work well for variable name, in ansible-freeipa we use `<what>_type`, in this case `grouping_type` or even `automember_type`. Also use this name as attribute in the YAML file. Please, do not use `type`.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:350]: ansible-freeipa modules must support `check_mode`. Check file [utils/templates/ipamodule+member.py.in](https://github.com/rafasgj/ansible-freeipa/blob/masterutils/templates/ipamodule+member.py.in), [line 289](https://github.com/freeipa/ansible-freeipa/blob/0cfd07a70966e5e43975c415be0445a774a1cf45/utils/templates/ipamodule%2Bmember.py.in#L289) to see how this is being implemented.

File: tests/automember/test_automember.yml
Comment (rjeffman) [inline tests/automember/test_automember.yml:22]: Please, and a testing group and hostgroup (with `ipagroup` and `ipahostgroup`), so that the whole group or hostgroup can be cleaned up before and after test is executed. (`testgroup` and `testhostgroup` are great names, in this case.)

**Review by rjeffman — CHANGES_REQUESTED** [review]

**Comment by rjeffman** [comment]
@jake2184 the failures on the checks, currently, have nothing to do with your changes as we are having some issues with the images used for testing. I'm fixing those, but we can't use the checks for a while. I'll let you know when everything is working again.

File: playbooks/automember/automember-absent.yml
Comment (jake2184) [reply playbooks/automember/automember-absent.yml:13]: Done

File: playbooks/automember/automember-present.yml
Comment (jake2184) [reply playbooks/automember/automember-present.yml:13]: Done

File: plugins/modules/ipaautomember.py
Comment (jake2184) [reply plugins/modules/ipaautomember.py:230]: Done

File: tests/automember/test_automember.yml
Comment (jake2184) [reply tests/automember/test_automember.yml:22]: Done

File: plugins/modules/ipaautomember.py
Comment (jake2184) [reply plugins/modules/ipaautomember.py:350]: Done

File: plugins/modules/ipaautomember.py
Comment (jake2184) [reply plugins/modules/ipaautomember.py:25]: Done

**Comment by rjeffman** [comment]
@chr15p, all test environments are up and running again. Please, fix the failing tests.

**Comment by rjeffman** [comment]
@chr15p, all test environments are up and running again. Please, fix the failing tests.

**Comment by jake2184** [comment]
Tests running cleanly

**Comment by rjeffman** [comment]
@jake2184 great job!

I'll try to do a proper review during this week.

**Comment by jake2184** [comment]
Can I bump this please @rjeffman ?

**Comment by jake2184** [comment]
And again @rjeffman @t-woerner

**Comment by rjeffman** [comment]
@jake2184 I was on vacation, and only returning "tomorrow".

I'll look at it Really Soon Now (tm). :-)

**Comment by jake2184** [comment]
No worries, as long as it doesn't fall through the cracks :)

**Comment by jake2184** [comment]
Is this going to be looked at any time soon? @rjeffman

**Comment by rjeffman** [comment]
Hello, sorry for the too long delay.

We're having issues with CI (and available time), as soon as CI is fixed, we'll restart pushing PRs, automount and automember modules are a priority.

I hope this to be soon, but can't promise as, right now, **time** is at a premium.

**Comment by rjeffman** [comment]
Hello @jake2184,

We had issues with CI, and the fixes require that all pending PR are rebased.

Can you please rebase your PR?

**Comment by jake2184** [comment]
Merged master in, pending CI checks

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:223]: Please, change `grouping` to `automember_type`. Try to use variable names as close as possible to the attribute names.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:174]: This function is not used anywhere.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:278]: This test is not needed as you will be using `inclusive_*`.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:293]: This test is not needed as you will be using `exclusive_*`.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:339]: For `state: rebuilt` the module should accept a list of user and host names.

File: tests/automember/test_automember.yml
Comment (rjeffman) [inline tests/automember/test_automember.yml:168]: Tests for `automember_type: hostgroup` with `action: member` are missing.

File: tests/automember/test_automember.yml
Comment (rjeffman) [inline tests/automember/test_automember.yml:107]: Tests for `automember_type: group` with `action: member` are missing.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@jake2184 thank you for the work so far.

There are still a few things to be fixed so that we can merge the automember module.

Check the code review comments, and, please, add example playbooks (`playbooks/automember`) with rules, as the ones currently available do not have any. Also add examples with `action: member` in the README, and in the playbooks directory.

**Comment by jake2184** [comment]
I've responded to most comments, awaiting CI outcome of the tests

File: plugins/modules/ipaautomember.py
Comment (jake2184) [reply plugins/modules/ipaautomember.py:330]: This is the only comment I don't think I've fixed.

To accept a list of objects, what are you proposing? Adding a new module argument just for this?

Comment (rjeffman) [reply plugins/modules/ipaautomember.py:330]: The CLI command automember-rebuild allows to rebuild automember rules for a list of users and/or hosts.

You should add twor parameters to the module `users` and `hosts` that are only valid when `state: rebuilt`. These parameters should be an optional list.

**Review by rjeffman — CHANGES_REQUESTED** [review]
Now the only thing missing is the parameters for `state: rebuilt`.

Nicely done!

File: tests/automember/test_automember.yml
Comment (rjeffman) [inline tests/automember/test_automember.yml:138]: This test is failing. Shouldn't it have `state: absent` ?

**Comment by jake2184** [comment]
Updated the tests, waiting for CI.

Also added "hosts" and "users" params to the rebuild module.

I'm not sure how multiple args should be passed to api.Command - I've just passed a list, will that work? 
``` 
                   commands.append([None, 'automember_rebuild',
                                    {"users": [
                                        to_text(_u)
                                        for _u in rebuild_users]}])

```

**Comment by t-woerner** [comment]
Yes, it should be like this.

**Comment by t-woerner** [comment]
test_automember.yml fails at the moment with 

TASK [Remove testgroup group automember rule member condition] *****************
task path: ansible-freeipa/tests/automember/test_automember.yml:138
fatal: [ipaserver.el83.local]: FAILED! => {"changed": false, "failed_when_result": true, "msg": "'dict' object has no attribute 'split'"}

**Comment by t-woerner** [comment]
`inclusive: [{'key': 'manager', 'expression': 'mscott'}]` is a list of dicts.

**Comment by jake2184** [comment]
Thanks for your pointers - though it is on me to fix the tests. I don't have a nice way to run them locally so am using the CI to run the tests, so don't necessarily expect them all to pass first time! I'll update here when I have them clean and in need of review.

**Comment by t-woerner** [comment]
I am using VMs to run the tests locally.

**Comment by t-woerner** [comment]
What do you think about creating a standalone PR for https://github.com/freeipa/ansible-freeipa/pull/486/commits/9ba44a3b1d9b23c9a9c10b47ec3261bcdb779661 ?
We could use it also in another change before this PR is ready to be merged.

**Comment by jake2184** [comment]
Well personally I'd much rather wait and have this merged. I've been waiting for months for this PR, we're proactively working on it, and it's very nearly there. Basically once tests are passing I don't think there is anything outstanding and it could be merged?

**Comment by t-woerner** [comment]
Ok, we will check the final PR again.

**Comment by t-woerner** [comment]
Please add 2021 to the date and add yourself to authors in ipaautomember.py

File: plugins/modules/ipaautomember.py
Comment (t-woerner) [inline plugins/modules/ipaautomember.py:5]: Please add yourself to Authors

File: plugins/modules/ipaautomember.py
Comment (t-woerner) [inline plugins/modules/ipaautomember.py:24]: Please import `ipalib_errors` from `ansible_freeipa_module` and use `ipalib_errors` as a replacement of `ipalib.errors`.

File: plugins/modules/ipaautomember.py
Comment (t-woerner) [inline plugins/modules/ipaautomember.py:130]: Please use `ipalib_errors` from `ansible_freeipa_module` instead.

File: plugins/modules/ipaautomember.py
Comment (t-woerner) [inline plugins/modules/ipaautomember.py:44]: Please put this in the same line as `description:` or use 
```
description: |
  Add, modify and delete an IPA Auto Membership Rules.
```

**Comment by jake2184** [comment]
Any idea why this is failing on Centos7 and not Centos8? Looks like the same python version, is the FreeIPA version the same? Can someone trigger a re-run of the failed test (I get auth denied)

File: plugins/module_utils/ansible_freeipa_module.py
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py:313]: It is better not to use ".keys()" here, please, change it to `[key for key in args if key not in ignore]`.

The main reason is that we are trying to enable pylint verification, and it wil fail if kept the way it is now.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:286]: I suggest simplifying these two lines into one: `key, regex = _inclusive.split("=", 1)`.

I also suggest to abort if key or regex is empty.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:294]: same as above.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:302]: same as above.

File: plugins/modules/ipaautomember.py
Comment (rjeffman) [inline plugins/modules/ipaautomember.py:310]: same as above.

**Comment by rjeffman** [comment]
Tested on the CLI (not the module) and it should work with Centos 7. Just restarted the test.

All FreeIPA version are different in the tested environments. CentOS 7 has FreeIPA version 4.6.8, CentOS 8 is 4.8.7 and Fedora is 4.9.3.

**Comment by rjeffman** [comment]
After re-running the tests, the same test still fails. Fixing it will require some debuging.

**Comment by rjeffman** [comment]
I could reproduce your issue locally, with CentOS-7, but have no time to debug it.

If you have `podman` installed, use [this script](https://gist.github.com/rjeffman/58d3989519e19af3f618b20e11f6057f) to test in a container and debug (usually, I debug with `ansible_module.warn("VAR: %s" % var)`).

**Comment by rjeffman** [comment]
For some reason, it looks like `key` and `automemberinclusiveregex` are swaped. I didn't look any further.

```
TASK [Add testhostgroup hostgroup automember rule member condition] ************************************
[WARNING]: testhostgroup automember_add_condition: {'automemberinclusiveregex': u'fqdn', 'type':
u'hostgroup', 'key': u'.*.domain.com'}
fatal: [dev-test-master]: FAILED! => {"changed": false, "failed_when_result": true, "msg": "automember_add_condition: testhostgroup: .*.domain.com is not a valid attribute."}
```

File: plugins/modules/ipaautomember.py
Comment (chr15p) [inline plugins/modules/ipaautomember.py:167]: This assumes that python lays out dicts in memory in a consistent fashion. It doesn't.
It assumes that the first item in the condition dict is key, and the second item is expression so condition.values gives [key, expression] and this outputs key=expression,  this isn't true on my RHEL7 box (and I assume Centos7)   
For me conditions is 
` [{'expression': '.*.domain.com', 'key': 'fqdn'}]`
so we end up with transformed returning  ".*.domain.com'=fqdn"  then later we assume that its key=value  and so the key and expression get swapped causing the testing errors you are seeing 

as a rough guess rewriting this as 
`transformed.append('%s=%s'%(condition['key'], condition['expression']) )`
would fix the issue   (no promises it wouldn't cause others elsewhere!)

It looks like the implementation details of different versions of python are breaking stuff for you.

**Comment by rjeffman** [comment]
@jake2184, sorry, but you'll need to rebase this PR.

For `plugins/module_utils/ansible_freeipa_module.py` just use what's on the `master` branch.

File: plugins/modules/ipaautomember.py
Comment (jake2184) [reply plugins/modules/ipaautomember.py:286]: huh. Never knew about that very handy parameter.

**Comment by jake2184** [comment]
Rebased, comments resolved, awaiting CI checks

**Review by rjeffman — APPROVED** [review]
LGTM.

**Comment by rjeffman** [comment]
@jake2184, code for automember is good.

Please, merge all the commits into a single one, and use a commit message base on the one found on PR #357.

File: README.md
Comment (t-woerner) [inline README.md:15]: Please also add the reference to README-automember.md in the end of the file.

File: tests/automember/test_automember.yml
Comment (t-woerner) [inline tests/automember/test_automember.yml:167]: Is this a comment?

File: README-automember.md
Comment (t-woerner) [inline README-automember.md:44]: The examples are not consistent. Please always add 
```
---
- name: Playbook to ...
  hosts: ipaserver
  become: yes
```

**Review by t-woerner — APPROVED** [review]
LGTM

**Comment by Wolskie** [comment]
@jake2184 thank you for taking over the work on this module, I appreciate it very much.

**Pull request merged**
